src/Avalonia.Controls/Presenters/TextPresenter.cs

```diff
@@ -969,5 +958,29 @@ protected override void OnPropertyChanged(AvaloniaPropertyChangedEventArgs chang
                     }
             }
         }
+
+        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
+        private static IImmutableBrush GetCaretBrush(IBrush? caretBrush, IBrush? background)
+        {
+            if (caretBrush is null)
+            {
+                var backgroundColor = (background as ISolidColorBrush)?.Color;
+
+                if (backgroundColor.HasValue)
+                {
+                    var red = (byte)~(backgroundColor.Value.R);
+                    var green = (byte)~(backgroundColor.Value.G);
+                    var blue = (byte)~(backgroundColor.Value.B);
+                    KnownColor known = (KnownColor)(0xFF << 24 | red << 16 | green << 8 | blue);
+
+                    caretBrush = KnownColors.ToBrush(known);
```

Comment:
This is bad practice. While an enum is technically allowed any value of the underlying type do NOT actually force any value into an enum.

The internal brush cash is very simple. Either it should: 1) Be extracted into another, internal 'BrushCache' class and shared or 2) The code re-implemented here.

Reply:
Thanks as always for the feedback.

These are considerations I have made. I asked myself a question. Why do I have to duplicate the code, increase the memory usage (even if only slightly) and reduce maintainability?

I answered myself that in this case maintaining good practice had more drawbacks than advantages.

I finally made this choice

Reply:
Also note:

 1. It is totally normal for types to have their own caching policy. That invovles some duplication. However, I would hardly say re-using a dictionary is much concern for duplication.
 2. Known colors at most has around ~150 values in the cache. Using the same cache here has 4_294_967_296 possible values... In practice it will be much, much less as not many CaretBrushes are ever created. Still there is a huge difference in magnitude in possible cache size. Especially for extremely long running apps that are changing this color for some reason.
 3. It is still possible to separate out this code into an internal `BrushCache` and share it everywhere. That could also allow setting a maximum size for the cache before older values are removed.